### ISSUE TITLE
Add boolean data type

### ISSUE THREAD

**Issue by emilsoman** [fika-lang/fika#1]
Fika currently doesn't have booleans, and that should change. To add booleans, we'll have to do these:

#### 1. Make the parser understand the strings "true" and "false" and create "boolean"-type nodes in the AST when they're read.

The goal is to make these tests pass:

```elixir
describe "boolean" do
  test "true" do
    str = "true"

    result = Parser.expression!(str)
    assert result ==  {:boolean, {1, 0, 4}, true}
  end

  test "false" do
    str = "false"

    result = Parser.expression!(str)
    assert result ==  {:boolean, {1, 0, 5}, false}
  end
end
```

Note that these tests don't currently exist, and need to be added to [test/fika/parser_test.exs](https://github.com/fika-lang/fika/blob/54e709e134da30eb93711da4aa551a8f60f26974/test/fika/parser_test.exs).

We'll look at how to make these tests pass. First, we'll have to open the file [lib/fika/parser.ex](https://github.com/fika-lang/fika/blob/54e709e134da30eb93711da4aa551a8f60f26974/lib/fika/parser.ex). What you see in this file is actually a parser combinator written using a library called [NimbleParsec](https://github.com/dashbitco/nimble_parsec). A parser combinator uses several smaller parsers to give you a big parser. In order to parse booleans, we're going to write a small parser for just that and plug it into the right place in the bigger parser. 

A parser for true | false looks like this:

```elixir
boolean =
  choice([
    string("true"),
    string("false")
  ])

```

Now if you look at other parsers like [the one for integer](https://github.com/fika-lang/fika/blob/54e709e134da30eb93711da4aa551a8f60f26974/lib/fika/parser.ex#L62-L65), you can see there's a final step of transforming the results into the AST using `Helper.to_ast` call. This generates the AST node that will represent what's parsed. So we'll add a new function in [lib/fika/parser_helper.ex](https://github.com/fika-lang/fika/blob/54e709e134da30eb93711da4aa551a8f60f26974/lib/fika/parser_helper.ex) that returns a node that looks like `{:boolean, line, true}`

This should be enough to pass the parser tests. We can run the tests using `mix test test/fika/parser_test.exs`.
The next step is to make the type checker aware of this new AST node.

#### 2. Type check booleans

Our goal is to make these tests pass:

```elixir
# These tests go to test/fika/type_checker_test.exs
describe "boolean" do
  test "true" do
    str = "true"
    ast = Fika.Parser.expression!(str)

    assert {:ok, "Bool", _} = TypeChecker.infer_exp(Env.init(), ast)
  end

  test "false" do
    str = "false"
    ast = Fika.Parser.expression!(str)

    assert {:ok, "Bool", _} = TypeChecker.infer_exp(Env.init(), ast)
  end
end
```

In order to do this, we'll open [the file for the type checker module](https://github.com/fika-lang/fika/blob/54e709e134da30eb93711da4aa551a8f60f26974/lib/fika/type_checker.ex)
and add a function that infers "Bool" types just like how the "Int" type is inferred.

#### 3. Translate to Erlang Abstract Format

In this step, we convert Fika AST to [Erlang Abstract Format](https://erlang.org/doc/apps/erts/absform.html).
Booleans are represented as `{:atom, line, true|false}` in Erlang abs form.
We need to make this test pass:

```elixir
# These tests go to test/fika/erl_translate_test.exs
test "boolean" do
  str = "true"
  ast = Parser.expression!(str)
  result = ErlTranslate.translate_expression(ast)

  assert result == {:atom, 1, true}
end
```

Open [lib/fika/erl_translate.ex](https://github.com/fika-lang/fika/blob/54e709e134da30eb93711da4aa551a8f60f26974/lib/fika/erl_translate.ex)
and add a function that converts boolean nodes to the format like the above.
Refer how integers are translated to get an idea.

#### 4. Run the code

If all went well, we should be able to write code with booleans. Let's try that.
Open `example.fi` and add a function that returns a boolean. Leave out the
return type for the function definition for now. Now open the elixir shell
using `iex -S mix` (`-S mix` loads the Fika project) and run `Fika.Code.load_file("example.fi")`
The type checker should catch the wrong return type. Let's fix that by specifying
Bool as the return type and reload the file. If the file was successfully compiled
and loaded, we can now call the function using `:example.function_name()` from
the shell.

#### 5. Update README

There's a section in the README that talks about data types in Fika and shows
examples. Update that part with the shiny new feature you just added. It's
time to get your pull request merged.

### Tasks
To recap, here are the tasks involved:

- [x] Update Parser
- [x] Update Type checker
- [x] Update Erl translator
- [x] Update README

All the best! If you need help, ping me in the discord server.

**Comment by emilsoman** [comment]
Update: This is reserved for the livestream happening on discord.

**Comment by emilsoman** [comment]
Closed by #6
